client/utils/auth.py

```diff
@@ -30,6 +28,25 @@
 </html>
 """
 
+def pick_free_port():
+    import socket
+    port_guess = 7777
+    while True:
+        try:
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.bind(('localhost', port_guess))
+        except Exception as e: # Something went wrong with the binding
+            print(e)
+            port_guess += 1
+            continue
+        finally:
+            s.close()
+        
+        return port_guess
+
+REDIRECT_PORT = pick_free_port()
```

Comment:
It's used in a couple of different places, not just in _make_code_post. You can also look at this as a global config variable since any communication with the student during auth will happen at localhost:REDIRECT_PORT.